Check chart urls are valid

brendarearden opened the pull request:
**Problem**
Some apps are attempted to be deployed with invalid urls which results in an error

**Solution**
Add an error on the catalog and forget the errors for invalid chart urls to ensure it does not keep refreshing through it

**Issue:**
https://github.com/rancher/rancher/issues/24914

**Dependent on:**
https://github.com/rancher/wrangler/pull/70
https://github.com/rancher/types/pull/1121
https://github.com/rancher/integration-test-charts/pull/12

cbron commented on pkg/catalog/manager/traverse.go:
```diff
@@ -273,6 +292,20 @@ func (m *Manager) traverseAndUpdate(helm *helmlib.Helm, commit string, cmt *Cata
+			}
+			logrus.Error(fmt.Sprintf("failed to sync templates. Multiple error(s) occured: %v", forgetErrors))
+			return &controller.ForgetError{Err: errors.Errorf(fmt.Sprintf("failed to sync templates. Multiple error(s) occured: %v", forgetErrors))}
+		}
```
So if there is a different error and an error here, this block would be skipped, is that ok ?

brendarearden replied:
yes because each `InvalidChartError` also gets added to the create and update errors, so they will get logged even if this portion gets skipped. It just ensures that if there is a weird error, we retry catalog refresh

cbron commented on pkg/catalog/manager/crd.go:
```diff
@@ -12,13 +13,23 @@ import (
 	TemplateNameLabel = "catalog.cattle.io/template_name"
 )
 
+type ForgettableError struct {
```
Interesting idea here. I don't think this name is most appropriate though. Something more like `InvalidChartError` ?

cbron commented on pkg/catalog/manager/traverse.go:
```diff
@@ -314,7 +347,43 @@ type catalogYml struct {
+	for _, url := range urls {
+		if validVersionURL(url) {
+			validURLs = append(validURLs, url)
+		}
```
Shouldn't you bubble up an error here ? I see part above that has if 0 valid urls create an error, but do they need to know which error ?

brendarearden replied:
https://github.com/rancher/rancher/blob/2e612b488c6ff6cb4bc3313411f34e16b641a0fb/pkg/catalog/manager/crd.go#L95. It gets bubbled here

cbron commented on tests/integration/suite/test_catalog.py:
```diff
@@ -313,11 +313,52 @@ def get_errored_catalog(catalog):
+    """Test chart with file:// and local:// url paths"""
+    client = admin_mc.client
+    name = random_str()
+    url = "https://github.com/brendarearden/integration-test-charts"
```
This will need to get switched to rancher

cbron commented on tests/integration/suite/test_catalog.py:
```diff
@@ -313,11 +313,52 @@ def get_errored_catalog(catalog):
+                       "catalog was not found in error state")
+    templates = client.list_template(catalogId=catalog.id).data
+
+    assert "aks-goodbyeworld" not in templates
```
Is this a bad url in the chart ?

brendarearden replied:
It is a bard url in the index.yaml of the chart repository - so its the url the deploy app is going to use to get the chart files - I added comments to help a little

cbron commented on pkg/catalog/manager/traverse.go:
```diff
@@ -314,7 +347,43 @@ type catalogYml struct {
 }
+
+func validVersionURL(versionURL string) bool {
+	if strings.HasPrefix(versionURL, "local://") || strings.HasPrefix(versionURL, "file://") {
```
Follow up here, this logic I linked is primarily for catalog not charts. Charts get deployed from app lifecycle, and they follow different pattern than catalogs. They could potentially use http/https/git protocols, so we can check for those specifically here or leave as-is.

cbron commented on pkg/catalog/manager/crd.go:
```diff
@@ -28,12 +39,21 @@ func (m *Manager) createTemplate(template v3.CatalogTemplate, catalog *v3.Catalo
 	for i := range template.Spec.Versions {
 		template.Spec.Versions[i].Files = nil
 		template.Spec.Versions[i].Readme = ""
+
```
nit: unecessary space

cbron commented on pkg/catalog/manager/traverse.go:
```diff
@@ -314,7 +348,47 @@ type catalogYml struct {
+	cmt.clusterCatalog = clusterCatalog
+}
+
+func onlyForgetErrors(forgetErrors, createErrors, updateErrors []error) bool {
```
forgot to change this name

rmweir commented on pkg/catalog/manager/crd.go:
```diff
@@ -72,6 +91,9 @@ func (m *Manager) updateTemplate(template *v3.CatalogTemplate, toUpdate v3.Catal
 		For each templateVersion in toUpdate, if spec doesn't match, do update
 		For version that doesn't exist, create a new one
 	*/
+	if len(toUpdate.Spec.Versions) == 0 || toUpdate.Spec.Versions == nil {
```
nit: could probably do this before grabbing and iterating template versions

rmweir commented on pkg/catalog/manager/catalog_sync.go:
```diff
@@ -37,6 +37,11 @@ func (m *Manager) Sync(key string, obj *v3.Catalog) (runtime.Object, error) {
 
+	// check to see if catalog was processed successfully, but had errors while processing
+	if v3.CatalogConditionProcessed.IsFalse(catalog) && v3.CatalogConditionRefreshed.IsTrue(catalog) {
+		return nil, nil
```
I think the comment above can be reworded, saying that it processed successfully but had errors is confusing. I also think there should be a comment explaining why you would return without an error despite process failing.

rmweir commented on pkg/catalog/manager/traverse.go:
```diff
@@ -273,6 +293,20 @@ func (m *Manager) traverseAndUpdate(helm *helmlib.Helm, commit string, cmt *Cata
+			if forgetErrorsLen < 5 {
+				maxTemplateNames = forgetErrorsLen
+			}
+			errorMessage := fmt.Sprintf("%s .... %v charts with invalid template url or name", strings.Join(errTemplateNames[:maxTemplateNames], " ; "), forgetErrorsLen)
```
maybe say actual number of templates failed, but 5 shown if forget is more than max

dramich commented on pkg/catalog/manager/catalog_sync.go:
```diff
@@ -37,6 +37,13 @@ func (m *Manager) Sync(key string, obj *v3.Catalog) (runtime.Object, error) {
+	// if the catalog was processed but some templates had errors due to local/file urls or chart names
+	// that cannot be used as labels - the catalog is up to date, but had errors so we don't refresh
+	// to give time for the user to make the necessary corrections.
+	if v3.CatalogConditionProcessed.IsFalse(catalog) && v3.CatalogConditionRefreshed.IsTrue(catalog) {
```
Is this needed? We should be processing the whole catalog and collecting errors then updating the status and commit. The commit check below should stop the catalog from coming through again

brendarearden replied:
not relying on commit to determine if refresh needs to occur again

dramich commented on pkg/catalog/manager/cluster_catalog_sync.go:
```diff
@@ -31,6 +31,13 @@ func (m *Manager) ClusterCatalogSync(key string, obj *v3.ClusterCatalog) (runtim
+	// if the catalog was processed but some templates had errors due to local/file urls or chart names
+	// that cannot be used as labels - the catalog is up to date, but had errors so we don't refresh
+	// to give time for the user to make the necessary corrections.
+	if v3.CatalogConditionProcessed.IsFalse(clusterCatalog) && v3.CatalogConditionRefreshed.IsTrue(clusterCatalog) {
```
Same comment as above

dramich commented on pkg/catalog/manager/crd.go:
```diff
@@ -12,14 +13,28 @@ import (
+
 func (m *Manager) createTemplate(template v3.CatalogTemplate, catalog *v3.Catalog) error {
+	// check to see if the template has valid urls - do not process if none exist
+	if len(template.Spec.Versions) == 0 || template.Spec.Versions == nil {
```
This check should be swapped around so the nil check is first.

brendarearden replied:
removed in favor of catalog preprocess

dramich commented on pkg/catalog/manager/crd.go:
```diff
@@ -12,14 +13,28 @@ import (
 func (m *Manager) createTemplate(template v3.CatalogTemplate, catalog *v3.Catalog) error {
+	// check to see if the template has valid urls - do not process if none exist
+	if len(template.Spec.Versions) == 0 || template.Spec.Versions == nil {
+		return &InvalidChartError{"must have one or more valid urls", template.Name}
```
The error doesn't convey what really happened, there isn't any data, not that it's invalid.

brendarearden replied:
removed as part of new implementation

dramich commented on pkg/catalog/manager/crd.go:
```diff
@@ -56,6 +76,10 @@ func (m *Manager) getTemplateMap(catalogName string, namespace string) (map[stri
 func (m *Manager) updateTemplate(template *v3.CatalogTemplate, toUpdate v3.CatalogTemplate) error {
+	// check to see if the version has valid urls - do not process if none exist
+	if len(toUpdate.Spec.Versions) == 0 || toUpdate.Spec.Versions == nil {
+		return &InvalidChartError{"must have one or more valid urls", template.Name}
```
The error doesn't convey what really happened, there isn't any data, not that it's invalid.

brendarearden replied:
removed as part of new implementation

brendarearden commented:
Changed to look for errors prior to processing catalog entries into templates. Added name and label processing in according with helm general conventions to ensure we are accepting what helm accepts.

brendarearden commented:
Per conversation with Dan I will be fixing the following items: 
1. remove the use of chart related naming conventions in favor of template and template version in the new pre process func 
2. do not truncate template name and template label strings and return error if larger than len 63 to keep inline with how it currently is
3. Return all errors in the UI instead of only truncated chart names to ensure all users can troubleshoot
4. Add url to the invalid url error message 
(if I missed any, feel free to add them)

dramich commented on pkg/catalog/manager/crd.go:
```diff
@@ -172,7 +168,7 @@ func (m *Manager) createTemplateVersions(catalogName string, versionsSpec []v3.T
 		templateVersion.Spec = spec
 		templateVersion.Status = v3.TemplateVersionStatus{HelmVersion: template.Status.HelmVersion}
-		templateVersion.Name = fmt.Sprintf("%s-%v", template.Name, spec.Version)
+		templateVersion.Name = getValidHelmChartLabel(template.Name, spec.Version)
```
Should this be passing through the label validation?

brendarearden replied:
I added label validation to the template name and remove it from the template version name

dramich commented on pkg/catalog/manager/traverse.go:
```diff
@@ -311,10 +324,19 @@ type catalogYml struct {
+}
+
+// Min returns the smaller of x or y.
+func Min(x, y int) int {
```
This doesn't appear to be used

dramich commented on pkg/catalog/manager/catalog_preprocess.go:
```diff
@@ -0,0 +1,92 @@
+			}
+			// validate label that follows general conventions and can be used as kubernetes label
+			templateVersionName := getValidTemplateNameWithVersion(templateName, version.Version)
+			errorString = validation.IsDNS1123Subdomain(templateVersionName)
```
Is this the correct check here? Should it also be IsQualifiedName? And the comment above this needs to be updated

brendarearden replied:
both checks need to be the same and they are not fully qualified

dramich commented on pkg/catalog/manager/traverse.go:
```diff
@@ -311,10 +324,19 @@ type catalogYml struct {
-	cmt.catalog = catalog
-	cmt.projectCatalog = projectCatalog
-	cmt.clusterCatalog = clusterCatalog
+func processInvalidChartErrors(entriesWithErrors []ChartsWithErrors) string {
```
Use a string builder with the loop, it will be cleaner/faster/less memory https://golang.org/pkg/strings/#Builder

brendarearden commented:
Also resolves https://github.com/rancher/rancher/issues/19279
